crates/red_knot_module_resolver/src/path.rs

```diff
@@ -1,676 +1,286 @@
-                        FilePathRef::System(path) => resolver.system().is_directory(path),
-                        FilePathRef::Vendored(path) => resolver.vendored().is_directory(path)
-                    }
-                }
+    pub(crate) fn push(&mut self, component: &str) {
+        if let Some(component_extension) = camino::Utf8Path::new(component).extension() {
+            assert!(
+                self.relative_path.extension().is_none(),
+                "Cannot push part {component} to {self:?}, which already has an extension"
+            );
+            if self.is_standard_library() {
+                assert_eq!(
+                    component_extension, "pyi",
+                    "Extension must be `pyi`; got `{component_extension}`"
+                );
+            } else {
+                assert!(
+                    matches!(component_extension, "pyi" | "py"),
+                    "Extension must be `py` or `pyi`; got `{component_extension}`"
+                );
             }
-            (path, root) => unreachable!(
-                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
-            )
         }
+        self.relative_path.push(component);
     }
 
     #[must_use]
-    fn is_regular_package(&self, search_path: Self, resolver: &ResolverState) -> bool {
-        fn is_non_stdlib_pkg(resolver: &ResolverState, path: &SystemPath) -> bool {
-            system_path_to_file(resolver.db.upcast(), path.join("__init__.py")).is_some()
-                || system_path_to_file(resolver.db.upcast(), path.join("__init__.py")).is_some()
-        }
-
-        match (self, search_path) {
-            (Self::Extra(path), Self::Extra(_)) => is_non_stdlib_pkg(resolver, path),
-            (Self::FirstParty(path), Self::FirstParty(_)) => is_non_stdlib_pkg(resolver, path),
-            (Self::SitePackages(path), Self::SitePackages(_)) => is_non_stdlib_pkg(resolver, path),
-            (Self::EditableInstall(path), Self::EditableInstall(_)) => is_non_stdlib_pkg(resolver, path),
-            // Unlike the other variants:
-            // (1) Account for VERSIONS
-            // (2) Only test for `__init__.pyi`, not `__init__.py`
-            (Self::StandardLibrary(path), Self::StandardLibrary(stdlib_root)) => {
-                match Self::query_stdlib_version( path, search_path, &stdlib_root, resolver) {
+    pub(crate) fn is_directory(&self, resolver: &ResolverState) -> bool {
+        let ModulePath {
+            search_path,
+            relative_path,
+        } = self;
+        match &*search_path.0 {
+            SearchPathInner::Extra(search_path)
+            | SearchPathInner::FirstParty(search_path)
+            | SearchPathInner::SitePackages(search_path)
+            | SearchPathInner::Editable(search_path) => resolver
+                .system()
+                .is_directory(&search_path.join(relative_path)),
```

Comment:
I don't think it's worth changing the design because of it. But the fact that the `ModulePath` is constructed out of the search path and a relative path requires that calling `is_directory` always allocates a new path (hidden behind `join`)

Reply:
Actually, almost all operations that now operate on the full path will require an allocation.

Reply:
I suppose we could store module paths as a `{search_path, absolute_path}` struct rather than a `{search_path, relative_path}` struct. But then we'd probably have to think about encoding the type in the `absolute_path` part of the struct again...

Reply:
Yeah, I think I would then go with the `kind` + `path` if we want to avoid that